### PR TITLE
Run Artifacts

### LINKED ISSUE #246: Add Runconfiguration Artifact definition
Allow exposing what artifacts a RunConfiguration/RunSchedule/Run can produce. This will allow dependency resolution and dependent triggering.

```yaml
kind: RunConfiguration
spec:
  run:
    ...
    artifacts:
    - name: a-long-artifact
      path: Pusher:pushed_model:0[pushed == 1]
    - name: a-short-artifact
      path: Pushed:pushed_model
  triggers:
    ...
```

Extend run completion events to include the resolved artifacts:

```yaml
{
  "id": "{{ UNIQUE_MESSAGE_ID }}",
  "specversion": "1.0",
  "source": "{{ PROVIDER_NAME }}",
  "type": "org.kubeflow.pipelines.run-completion",
  "datacontenttype": "application/json",
  "data": {
    "status": "succeeded|failed",
    "pipelineName":"{{ PIPELINE_NAME }}",
    "runConfigurationName": "{{ RUNCONFIGURATION_NAMESPACE }}/{{ RUNCONFIGURATION_NAME }}", 
    "runName": "{{ RUN_NAMESPACE }}/{{ RUN_NAME }}", 
    "runId": "{{ RUN_ID }}",
    "servingModelArtifacts": [
      {
        "name":"{{ PIPELINE_NAME }}:{{ WORKFLOW_NAME }}:Pusher:pushed_model:{{ PUSHER_INDEX }}",
        "location":"gs://{{ PIPELINE_ROOT }}/Pusher/pushed_model/{{ MODEL_VERSION }}"
      }
    ],
    "artifacts": [
      {"name": "{{ ARTIFACT_NAME}} ", "location": "{{ ARTIFACT_LOCATION}}"}
    ]
  }
}
```

Notes:
- When index is omitted, default to '0'
- Wildcards are out of scope
- Consider using https://github.com/hashicorp/go-bexpr for evaluating boolean expression filters
- `servingModelArtifacts` in run completion events will be deprecated but not removed

Discarded Options:

```yaml
kind: RunConfiguration
spec:
  artifacts:
  - name: a-long-artifact
    component: Pusher
    artifact: pushed_model
    filter:
    - key: pushed
      value: 1
  - name: a-short-artifact
    artifact: pushed_model
```

### PR DESCRIPTION
**Pull request by jmendesky**
Closes #246 

## Tasks
- [x] Add OutputArtifact Definition
- [x] Resolve Artifacts in eventing servers
- [x] Documentation updated (including events)

Note: this alters versions of runschedules and runs because a new field has been introduced

### CODE REVIEW

File: apis/pipelines/v1alpha5/artifact_test.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+import (
+	"fmt"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/sky-uk/kfp-operator/apis"
+	"math/rand"
+)
+
+var _ = Context("RunConfiguration", func() {
```
Comment (jmendesky) [inline apis/pipelines/v1alpha5/artifact_test.go:14]: Rename to Artifact

File: apis/pipelines/v1alpha5/runconfiguration_types.go [hunk 2]
```diff
@@ -21,8 +21,8 @@ var OnChangeTypes = struct {
 }
 
 type RunConfigurationSpec struct {
-	Run      RunSpec  `json:"run,omitempty"`
-	Triggers Triggers `json:"triggers,omitempty"`
+	Run       RunSpec    `json:"run,omitempty"`
```
Comment (jmendesky) [inline apis/pipelines/v1alpha5/runconfiguration_types.go:24]: Revert formatting

File: apis/pipelines/v1alpha5/artifact.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+func (ap *ArtifactPath) UnmarshalText(bytes []byte) error {
+	deserialised, err := ArtifactPathFromString(string(bytes))
+	*ap = deserialised
+
+	return err
+}
+
+type Artifact struct {
```
Comment (jmendesky) [inline apis/pipelines/v1alpha5/artifact.go:84]: Rename to ArtifactSpec?

File: argo/providers/kfp/metadata_store.go [hunk 4]
```diff
@@ -58,21 +49,60 @@ func (gms *GrpcMetadataStore) GetServingModelArtifact(ctx context.Context, workf
...
-					Name:     artifactName,
-					Location: artifactUri,
-				})
+		artifactUri := artifact.GetUri()
+		artifactName := artifact.GetCustomProperties()[ArtifactNameCustomProperty].GetStringValue()
+		if artifactUri == "" {
+			continue
+		}
+
+		for _, artifactDef := range artifactDefs {
```
Comment (jmendesky) [inline argo/providers/kfp/metadata_store.go:60]: move this loop outside to reduce the number of time evaluator needs to be created

File: argo/providers/kfp/metadata_store.go [hunk 5]
```diff
@@ -58,21 +49,60 @@ func (gms *GrpcMetadataStore) GetServingModelArtifact(ctx context.Context, workf
...
 			}
+
+			if artifactDef.Path.Filter != "" {
+				evaluator, err := bexpr.CreateEvaluator(artifactDef.Path.Filter)
+				if err != nil {
+					return nil, err
+				}
+
+				matched, err := evaluator.Evaluate(propertiesToPrimitiveMap(artifact.GetCustomProperties()))
+				if err != nil {
```
Comment (jmendesky) [inline argo/providers/kfp/metadata_store.go:72]: comment why we are not propagating the error here

File: argo/providers/vai/eventing_server.go [hunk 6]
```diff
@@ -183,11 +140,49 @@ func toRunCompletionEvent(job *aiplatformpb.PipelineJob, runId string) *common.R
...
+	return artifactsForJob(job, []pipelinesv1.Artifact{base.LegacyArtifactDefinition})
+}
+
+func artifactsForJob(job *aiplatformpb.PipelineJob, artifactDefs []pipelinesv1.Artifact) []common.Artifact {
+	var servingModelArtifacts []common.Artifact
+	for _, task := range job.GetJobDetail().GetTaskDetails() {
+		for outputName, output := range task.GetOutputs() {
+			for _, artifact := range output.GetArtifacts() {
+				for _, artifactDef := range artifactDefs {
+					if task.TaskName == artifactDef.Path.Locator.Component && outputName == artifactDef.Path.Locator.Artifact && artifact.Uri != "" {
```
Comment (jmendesky) [inline argo/providers/vai/eventing_server.go:159]: No use of `index`

File: apis/pipelines/v1alpha5/run_types.go [hunk 7]
```diff
@@ -12,6 +12,7 @@ type RunSpec struct {
 	Pipeline          PipelineIdentifier `json:"pipeline,omitempty"`
 	ExperimentName    string             `json:"experimentName,omitempty"`
 	RuntimeParameters []apis.NamedValue  `json:"runtimeParameters,omitempty"`
+	Artifacts         []OutputArtifact   `json:"artifacts,omitempty"`
```
Comment (jmendesky) [inline apis/pipelines/v1alpha5/run_types.go:15]: This should change the run's computed hash

File: docs-gen/content/docs/reference/resources/runconfiguration.md [hunk 8]
```diff
@@ -33,6 +36,7 @@ A Run Configuration can have one of more triggers that determine when the next t
 |------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
 | `spec.run.pipeline`          | The [identifier](../pipeline/#identifier) of the corresponding pipeline resource to run. If no version is specified, then the RunConfiguration will track the latest version of the specified pipeline.                                           |
 | `spec.run.experimentName`    | The name of the corresponding experiment resource (optional - the `Default` Experiment as defined in the [Installation and Configuration section of the documentation](README.md#configuration) will be used if no `experimentName` is provided). |
+| `spec.run.artifacts[]`       | Exposed models that will be included in run completion event when a run for this RunConfiguration has succeeded. See [Run Artifact Definition](../run/#run-artifact-definition) for more information.                                             |
```
Comment (jmendesky) [inline docs-gen/content/docs/reference/resources/runconfiguration.md:39]:
```suggestion
| `spec.run.artifacts[]`       | Exposed output artifacts that will be included in run completion events when a run for this RunConfiguration has succeeded. See [Run Artifact Definition](../run/#run-artifact-definition) for more information.                                             | 
```

File: docs-gen/content/docs/reference/resources/run.md [hunk 9]
```diff
@@ -28,6 +31,25 @@ Note the usage of `metadata.generateName` which tells Kubernetes to generate a n
 | `spec.pipeline`          | The [identifier](../pipeline/#identifier) of the corresponding pipeline resource to run. If no version is specified, then the RunConfiguration will use the latest version of the specified pipeline.                                             |
 | `spec.experimentName`    | The name of the corresponding experiment resource (optional - the `Default` Experiment as defined in the [Installation and Configuration section of the documentation](README.md#configuration) will be used if no `experimentName` is provided). |
 | `spec.runtimeParameters` | Dictionary of runtime parameters as exposed by the pipeline.                                                                                                                                                                                      |
+| `spec.run.artifacts[]`   | Exposed artifacts that will be included in run completion event when this run has succeeded. See below for more information.                                                                                                                      |
```
Comment (jmendesky) [inline docs-gen/content/docs/reference/resources/run.md:34]:
```suggestion
| `spec.run.artifacts[]`   | Exposed output artifacts that will be included in run completion event when this run has succeeded. See below for more information.                                                                                                                      |
```

File: argo/providers/kfp/metadata_store.go [hunk 10]
```diff
@@ -76,3 +80,79 @@ func (gms *GrpcMetadataStore) GetServingModelArtifact(ctx context.Context, workf
...
+				return nil, err
+			}
+		}
+
+		for _, artifact := range artifactsResponse.GetArtifacts() {
+			artifactUri := artifact.GetUri()
+			if artifactUri == "" {
+				continue
+			}
+			if !strings.HasSuffix(artifact.GetName(), artifactDef.Path.Locator.String()) {
```
Comment (jmendesky) [inline argo/providers/kfp/metadata_store.go:116]: should this be `displayName`?

File: argo/providers/kfp/metadata_store_unit_test.go [hunk 11]
```diff
@@ -35,7 +38,133 @@ var _ = Context("gRPC Metadata Store", func() {
...
+		When("GetArtifactsByContext returns artifacts", func() {
+			It("filters out all invalid artifacts", func() {
+				contextId := givenContextId()
+				artifactLocation := common.RandomString()
+				mockMetadataStoreServiceClient.EXPECT().
+					GetArtifactsByContext(gomock.Any(), gomock.Eq(&ml_metadata.GetArtifactsByContextRequest{ContextId: &contextId})).
+					Return(&ml_metadata.GetArtifactsByContextResponse{
+						Artifacts: []*ml_metadata.Artifact{
+							{
+								Name: pointer.String(common.RandomString()),
```
Comment (jmendesky) [inline argo/providers/kfp/metadata_store_unit_test.go:119]: add case for missing URI and use different URI for each

**Comment by jmendesky** [comment]
ok to test

File: apis/pipelines/v1alpha5/artifact.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+	"fmt"
+	"regexp"
+	"strconv"
+)
+
+type ArtifactLocator struct {
+	Component string `json:"-" yaml:"-"`
+	Artifact  string `json:"-" yaml:"-"`
+	Index     int    `json:"-" yaml:"-"`
+}
```
Comment (alexgeorgousis) [inline apis/pipelines/v1alpha5/artifact.go:13]: What does the `-` syntax do?

File: apis/pipelines/v1alpha5/artifact.go [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+func (ap ArtifactPath) String() string {
+	if ap.Filter == "" {
+		return ap.Locator.String()
+	}
+
+	return fmt.Sprintf("%s[%s]", ap.Locator.String(), ap.Filter)
+}
+
+func ArtifactPathFromString(path string) (artifactPath ArtifactPath, err error) {
```
Comment (alexgeorgousis) [inline apis/pipelines/v1alpha5/artifact.go:36]: I assume we can't have anything like a static method in Go

File: apis/pipelines/v1alpha5/artifact.go [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+	pathPattern := regexp.MustCompile(ArtifactPathPattern)
+	matches := pathPattern.FindStringSubmatch(path)
+
+	if len(matches) < 3 {
+		err = fmt.Errorf("ArtifactPath must match pattern %s", ArtifactPathPattern)
+		return
```
Comment (alexgeorgousis) [inline apis/pipelines/v1alpha5/artifact.go:42]: I'm a bit confused by these empty returns

**Review by alexgeorgousis — APPROVED** [review]

**Pull request merged**